File: algorithms/java/src/test/java/leetcode/LC049GroupAnagrams.java
package leetcode.lc049_group_anagrams;

import java.util.*;

import org.junit.jupiter.api.Test;


import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * https://leetcode.com/problems/group-anagrams/
 * #medium
 */
public final class LC049GroupAnagrams {
    public final class Solution {
        public List<List<String>> groupAnagrams(final String[] strs) {
            Map<String, List<String>> seen = new HashMap<>();
            for (String str : strs) {
                char[] array = str.toCharArray();
                Arrays.sort(array);
                String key = new String(array);
                if (seen.containsKey(key)) {
                    seen.get(key).add(str);
                } else {
                    List<String> group = new ArrayList<>();
                    group.add(str);
                    seen.put(key, group);
                }
            }
            ArrayList<List<String>> result = new ArrayList<>();
            for (List<String> group : seen.values()) {
                Collections.sort(group);
                result.add(group);
            }
            return result;
        }
    }

    private final class OrderlyComparator implements Comparator<List<String>>, Serializable {
        @Override
        public int compare(final List<String> l1, final List<String> l2) {
            int difference = l1.size() - l2.size();
            if (0 != difference) {
                return difference;
            } else {
                for (int i = 0; i < l1.size(); i++) {
                    difference = l1.get(i).compareTo(l2.get(i));
                    if (0 != difference) {
                        return difference;
                    }
                }
                return 0;
            }
        }
    }

    public void test(final String[][] expected, final List<List<String>> result) throws Exception {
        Collections.sort(result, new OrderlyComparator());
        assertEquals(expected.length, result.size());
        for (int i = 0; i < expected.length; i++) {
            assertEquals(expected[i].length, result.get(i).size());
            for (int j = 0; j < expected[i].length; j++) {
                assertEquals(expected[i][j], result.get(i).get(j));
            }
        }
    }

    @Test
    public void test_abc_cab_bad_dab_zzz_dot() throws Exception {
        String[] strs = {"abc", "cab", "bad", "dab", "zzz", "dot"};
        String[][] expected = {{"dot"}, {"zzz"}, {"abc", "cab"}, {"bad", "dab"}};
        List<List<String>> result = new Solution().groupAnagrams(strs);
        test(expected, result);
    }

    @Test
    public void test_tea_and_ate_eat_den() throws Exception {
        String[] strs = {"tea", "and", "ate", "eat", "den"};
        String[][] expected = {{"and"}, {"den"}, {"ate", "eat", "tea"}};
        List<List<String>> result = new Solution().groupAnagrams(strs);
        test(expected, result);
    }
}

File: algorithms/code/leetcode/lc049_group_anagrams/lc049_group_anagrams.py
#!/usr/bin/env python3
# https://leetcode.com/problems/group-anagrams/

import unittest
from typing import Any, Callable


class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        seen: dict[str, list[str]] = {}
        for string in strs:
            array = list(string)
            array.sort()
            key = ''.join(array)
            if key in seen:
                seen[key].append(string)
            else:
                group = []
                group.append(string)
                seen[key] = group
        result = []
        for group in seen.values():
            group.sort()
            result.append(group)
        return result


class TestCode(unittest.TestCase):
    @staticmethod
    def cmp_to_key(mycmp: Callable) -> Any:  # pragma: no cover
        class K:
            def __init__(self, obj: Any, *_: Any) -> None:
                self.obj = obj

            def __lt__(self, other: Any) -> bool:
                return mycmp(self.obj, other.obj) < 0

            def __gt__(self, other: Any) -> bool:
                return mycmp(self.obj, other.obj) > 0

            def __eq__(self, other: Any) -> bool:
                return mycmp(self.obj, other.obj) == 0

            def __le__(self, other: Any) -> bool:
                return mycmp(self.obj, other.obj) <= 0

            def __ge__(self, other: Any) -> bool:
                return mycmp(self.obj, other.obj) >= 0

            def __ne__(self, other: Any) -> bool:
                return mycmp(self.obj, other.obj) != 0

        return K

    @staticmethod
    def orderly_comparator(
        l1: list[str], l2: list[str]
    ) -> int:  # pragma: no cover
        difference = len(l1) - len(l2)
        if difference != 0:
            return difference
        for l1s, l2s in zip(l1, l2, strict=True):
            if l1s < l2s:
                return -1
            if l1s > l2s:
                return 1
        return 0

    def __test(
        self, expected: list[list[str]], result: list[list[str]]
    ) -> None:
        result = sorted(
            result, key=TestCode.cmp_to_key(TestCode.orderly_comparator)
        )
        self.assertEqual(len(expected), len(result))
        for i, expected_i in enumerate(expected):
            self.assertEqual(len(expected_i), len(result[i]))
            for j, expected_i_j in enumerate(expected_i):
                self.assertEqual(expected_i_j, result[i][j])

    def test_abc_cab_bad_dab_zzz_dot(self) -> None:
        strs = ['abc', 'cab', 'bad', 'dab', 'zzz', 'dot']
        expected = [['dot'], ['zzz'], ['abc', 'cab'], ['bad', 'dab']]
        result = Solution().groupAnagrams(strs)
        self.__test(expected, result)

    def test_tea_and_ate_eat_den(self) -> None:
        strs = ['tea', 'and', 'ate', 'eat', 'den']
        expected = [['and'], ['den'], ['ate', 'eat', 'tea']]
        result = Solution().groupAnagrams(strs)
        self.__test(expected, result)
